crawler/main.py

```diff
@@ -19,71 +26,117 @@
+            )
+            samples_collection.create_index(FIELD_PLATE_BARCODE)
+            logger.debug(f"Creating compund index on '{samples_collection.full_name}'")
+            # create compound index on 'Root Sample ID', 'RNA ID', 'Result', 'Lab ID' - some data
+            #   had the same plate tested at another time so ignore the data if it is exactly the
+            #   same
+            samples_collection.create_index(
+                [
+                    (FIELD_ROOT_SAMPLE_ID, pymongo.ASCENDING),
+                    (FIELD_RNA_ID, pymongo.ASCENDING),
+                    (FIELD_RESULT, pymongo.ASCENDING),
+                    (FIELD_RNA_ID, pymongo.ASCENDING),
+                    (FIELD_LAB_ID, pymongo.ASCENDING),
+                ],
+                unique=True,
+            )
+
             for centre in centres:
                 logger.debug(f"Processing {centre['name']}")
 
+                docs_inserted: int = 0
+                latest_file_name: str = ""
+                errors: List[str] = []
                 try:
                     if sftp:
-                        download_csv(config, centre)
+                        download_csv_files(config, centre)
+
+                    if "merge_required" in centre.keys() and centre["merge_required"]:
+                        master_file_name = merge_daily_files(centre)
+
+                        # only upload to SFTP if config explicitly says so - this is to prevent
+                        #   accidental uploads from non-production envs
+                        upload = config.SFTP_UPLOAD  # type: ignore
+                        if upload:
+                            upload_file_to_sftp(config, centre, master_file_name)
 
                     latest_file_name, errors, docs_to_insert = parse_csv(centre)
 
                     logger.debug(f"Attempting to insert {len(docs_to_insert)} docs")
-                    result = samples_collection.insert_many(docs_to_insert)
-                    logger.debug(f"{len(result.inserted_ids)} documents inserted")
+                    result = samples_collection.insert_many(docs_to_insert, ordered=False)
+
+                    docs_inserted = len(result.inserted_ids)
+                except BulkWriteError as e:
+                    # This is happening when there are duplicates in the data and the index prevents
+                    #   the records from being written
+                    logger.warning(e)
+                    docs_inserted = e.details["nInserted"]
+                    write_errors = {write_error["code"] for write_error in e.details["writeErrors"]}
+                    for error in write_errors:
+                        num_errors = len(
+                            list(filter(lambda x: x["code"] == error, e.details["writeErrors"]))
+                        )
+                        errors.append(f"{num_errors} records with error code {error}")
+                except Exception as e:
```

Comment:
As mentioned in slack, this is where I think I'd either expect to see handling of the CentreFileError exception, or possibly just populating the errors list with the more general exceptions (I'm not sure what else ends up in here). Again, may easily be missing something.